sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+    dif_padctrl_attr_t bad_attrs_1 = dif_padctrl_attr_t(
+        (1 << PADCTRL_PARAM_ATTRDW) | kDifPadctrlAttrOpenDrain);
+
+    EXPECT_EQ(
+        dif_padctrl_disable_attrs(&dif_padctrl_, pad_kind, 0, bad_attrs_1),
+        kDifPadctrlChangeBadArg);
+
+    dif_padctrl_attr_t bad_attrs_2 = dif_padctrl_attr_t(-1);
+
+    EXPECT_EQ(
+        dif_padctrl_disable_attrs(&dif_padctrl_, pad_kind, 0, bad_attrs_2),
+        kDifPadctrlChangeBadArg);
+  }
+}
+
+TEST_F(DifPadctrlTest, WriteAttrsFailureInvalidAttr) {
+  DeviceInit();
+
+  for (auto pad_kind : {kDifPadctrlPadMio, kDifPadctrlPadDio}) {
+    dif_padctrl_attr_t bad_attrs_1 = dif_padctrl_attr_t(
+        (1 << PADCTRL_PARAM_ATTRDW) | kDifPadctrlAttrOpenDrain);
+
+    EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind, 0, bad_attrs_1),
+              kDifPadctrlChangeBadArg);
+
+    dif_padctrl_attr_t bad_attrs_2 = dif_padctrl_attr_t(-1);
+
+    EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind, 0, bad_attrs_2),
+              kDifPadctrlChangeBadArg);
+  }
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_write_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, GetAttrsSuccess) {
+  DeviceInit();
+
+  for (auto attr_field : mio_attr_fields_) {
+    dif_padctrl_attr_t expected_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrPullDown | kDifPadctrlAttrIOInvert);
+    ExpectAttrFieldRead(attr_field, expected_attrs);
+
+    dif_padctrl_attr_t seen_attrs = kDifPadctrlAttrNone;
```

Comment:
Is there a reason this value needs to be initialized? You don't initialize it for other calls of the same function in other tests.